activesupport/lib/active_support/rescuable.rb

```diff
@@ -65,65 +63,132 @@ def rescue_from(*klasses, &block)
+      # raising a generic outer error to indicate one of many internal
+      # errors, the outer exception's cause is set to the original exception:
+      #
+      #     begin
+      #       raise original
+      #     rescue
+      #       raise wrapped
+      #     end
+      #
+      # Since we raised wrapped while handling original,
+      #
+      #     wrapped.cause == original
+      #
+      # When an exception with a #cause is handled, we use the highest
+      # priority handler for both the exception and its cause. That means
+      # that a handler declared for one exception class may be called with
+      # an exception of a different class (the cause):
+      #
+      #     rescue_from WrappedException, with: …
+      #     rescue_from OriginalException, with: …
+      #
+      # The last handler declaration is highest priority, so raising a
+      # wrapped exception with an original exception as its cause will be
+      # handled by the OriginalException handler, not the WrappedException
+      # handler.
+      #
+      # The aim here is to provide generic exception handling that can be
+      # overridden with handling for more specific causes, like a generic
+      # Active Record error handling and a specific PostgreSQL unique
+      # constraint violation error.
+      def rescue_with_handler(exception, object: self, and_cause: false)
+        if handler = handler_for_rescue(exception, object: object, and_cause: and_cause)
+          handler.arity != 0 ? handler.call(exception) : handler.call
+          true # don't rely on the return value of the handler
+        end
       end
-    end
 
-    def handler_for_rescue(exception)
-      # We go from right to left because pairs are pushed onto rescue_handlers
-      # as rescue_from declarations are found.
-      _, rescuer = self.class.rescue_handlers.reverse.detect do |klass_name, handler|
-        # The purpose of allowing strings in rescue_from is to support the
-        # declaration of handler associations for exception classes whose
-        # definition is yet unknown.
-        #
-        # Since this loop needs the constants it would be inconsistent to
-        # assume they should exist at this point. An early raised exception
-        # could trigger some other handler and the array could include
-        # precisely a string whose corresponding constant has not yet been
-        # seen. This is why we are tolerant to unknown constants.
-        #
-        # Note that this tolerance only matters if the exception was given as
-        # a string, otherwise a NameError will be raised by the interpreter
-        # itself when rescue_from CONSTANT is executed.
-        klass = self.class.const_get(klass_name) rescue nil
-        klass ||= (klass_name.constantize rescue nil)
-        klass === exception if klass
+      def handler_for_rescue(exception, object: self, and_cause: false) #:nodoc:
+        case rescuer = rescuer_for_exception(exception, and_cause: and_cause)
```

Comment:
Extract rescuer lookup which had been duplicated in `index_of_handler_for_rescue` utility method.